File: Photon-Tinker/ScanBarcodeViewController.m
//
//  igViewController.m
//  ScanBarCodes
//
//  Created by Torrey Betts on 10/10/13.
//  Copyright (c) 2013 Infragistics. All rights reserved.
//

#import <AVFoundation/AVFoundation.h>
#import "ScanBarcodeViewController.h"

@interface ScanBarcodeViewController () <AVCaptureMetadataOutputObjectsDelegate>
{
    AVCaptureSession *_session;
    AVCaptureDevice *_device;
    AVCaptureDeviceInput *_input;
    AVCaptureMetadataOutput *_output;
    AVCaptureVideoPreviewLayer *_prevLayer;

    UIView *_highlightView;
    UILabel *_label;
    UIImageView *_overlayImageView;
    UIButton *_overlayButton;
    UIButton *_cancelButton;
    UIView *_circleView;
}
@end

@implementation ScanBarcodeViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    _highlightView = [[UIView alloc] init];
    //_highlightView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin|UIViewAutoresizingFlexibleLeftMargin|UIViewAutoresizingFlexibleRightMargin|UIViewAutoresizingFlexibleBottomMargin;
    _highlightView.layer.borderColor = [UIColor greenColor].CGColor;
    _highlightView.layer.borderWidth = 3;
    [self.view addSubview:_highlightView];

    _label = [[UILabel alloc] init];
    _label.frame = CGRectMake(0, self.view.bounds.size.height - 40, self.view.bounds.size.width, 40);
    //_label.autoresizingMask = UIViewAutoresizingFlexibleTopMargin;
    _label.backgroundColor = [UIColor colorWithWhite:0.15 alpha:0.65];
    _label.textColor = [UIColor whiteColor];
    _label.textAlignment = NSTextAlignmentCenter;
    _label.text = @"Point at SIM ICCID barcode";
    [self.view addSubview:_label];

    _overlayImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"overlaygraphic"]];
    [_overlayImageView setFrame:CGRectMake(30, 150, self.view.bounds.size.width-60, self.view.bounds.size.height-300)];
    _overlayImageView.image = [_overlayImageView.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    [_overlayImageView setTintColor:[UIColor whiteColor]];
    [self.view addSubview:_overlayImageView];
    
    _overlayButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [_overlayButton setImage:[UIImage imageNamed:@"flashlight"] forState:UIControlStateNormal];
    [_overlayButton setFrame:CGRectMake(30, 30, 48, 48)];
    [_overlayButton setTintColor:[UIColor whiteColor]];
    [_overlayButton addTarget:self action:@selector(toggleTorchButton:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_overlayButton];

    _cancelButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [_cancelButton setTitle:@"Cancel" forState:UIControlStateNormal];
    [_cancelButton setFrame:CGRectMake(self.view.bounds.size.width-60, self.view.bounds.size.height-44, 60, 48)];
    [_cancelButton addTarget:self action:@selector(cancelScan:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_cancelButton];

    
    _circleView = [[UIView alloc] initWithFrame:CGRectMake(22, 24, 64, 64)];
    _circleView.alpha = 0.5;
    _circleView.layer.cornerRadius = 32;
    _circleView.backgroundColor = [UIColor grayColor];
    [self.view addSubview:_circleView];
    
    _session = [[AVCaptureSession alloc] init];
    _device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    if ([_device hasTorch]) {
        [_device lockForConfiguration:nil];
        [_device setTorchMode:AVCaptureTorchModeOn];  // use AVCaptureTorchModeOff to turn off
        [_device unlockForConfiguration];
    }
    NSError *error = nil;

    _input = [AVCaptureDeviceInput deviceInputWithDevice:_device error:&error];
    if (_input) {
        [_session addInput:_input];
    } else {
        NSLog(@"Error: %@", error);
    }

    _output = [[AVCaptureMetadataOutput alloc] init];
    [_output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    [_session addOutput:_output];

    _output.metadataObjectTypes = [_output availableMetadataObjectTypes];

    _prevLayer = [AVCaptureVideoPreviewLayer layerWithSession:_session];
    _prevLayer.frame = self.view.bounds;
    _prevLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    [self.view.layer addSublayer:_prevLayer];

    [_session startRunning];

    [self.view bringSubviewToFront:_highlightView];
    [self.view bringSubviewToFront:_label];
    [self.view bringSubviewToFront:_circleView];
    [self.view bringSubviewToFront:_overlayButton];
    [self.view bringSubviewToFront:_overlayImageView];
    [self.view bringSubviewToFront:_cancelButton];

    
}


- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];

    if (@available(iOS 11.0, *)) {
        UIEdgeInsets insets = self.view.safeAreaInsets;


        _label.frame = CGRectMake(0, self.view.bounds.size.height - 40 - insets.bottom, self.view.bounds.size.width, 40);
        _cancelButton.frame = CGRectMake(self.view.bounds.size.width-60, self.view.bounds.size.height - 44 - insets.bottom, 60, 48);

        _overlayButton.frame = CGRectMake(30, 30 + insets.top, 48, 48);
        _circleView.frame = CGRectMake(22, 24 + insets.top, 64, 64);

        _overlayImageView.frame = CGRectMake(30, 150 + insets.top, self.view.bounds.size.width-60, self.view.bounds.size.height - 300 - insets.top - insets.bottom);
    }
}


- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleLightContent;
}


-(void)toggleTorchButton:(id)sender
{
    if ([_device hasTorch]) {
        [_device lockForConfiguration:nil];
        if (_device.torchMode == AVCaptureTorchModeOff)
            [_device setTorchMode:AVCaptureTorchModeOn];
        else
            [_device setTorchMode:AVCaptureTorchModeOff];
        [_device unlockForConfiguration];
    }

}




-(void)cancelScan:(id)sender
{
    [self.delegate didCancelScanningBarcode:self];
}


- (IBAction)test:(id)sender {
    [_session stopRunning];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.55 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self.delegate didFinishScanningBarcodeWithResult:self barcodeValue:@"1234567890123456789012"];
    });

}


- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection
{
    CGRect highlightViewRect = CGRectZero;
    AVMetadataMachineReadableCodeObject *barCodeObject;
    NSString *detectionString = nil;
//    NSArray *barCodeTypes = @[AVMetadataObjectTypeUPCECode, AVMetadataObjectTypeCode39Code, AVMetadataObjectTypeCode39Mod43Code,
//            AVMetadataObjectTypeEAN13Code, AVMetadataObjectTypeEAN8Code, AVMetadataObjectTypeCode93Code, AVMetadataObjectTypeCode128Code,
//            AVMetadataObjectTypePDF417Code, AVMetadataObjectTypeQRCode, AVMetadataObjectTypeAztecCode];

    NSArray *barCodeTypes = @[AVMetadataObjectTypeCode128Code]; //ido

    for (AVMetadataObject *metadata in metadataObjects) {
        for (NSString *type in barCodeTypes) {
            if ([metadata.type isEqualToString:type])
            {
                barCodeObject = (AVMetadataMachineReadableCodeObject *)[_prevLayer transformedMetadataObjectForMetadataObject:(AVMetadataMachineReadableCodeObject *)metadata];
                highlightViewRect = barCodeObject.bounds;
                detectionString = [(AVMetadataMachineReadableCodeObject *)metadata stringValue];
                break;
            }
        }

        if (detectionString != nil)
        {
            _label.text = detectionString;
//            [_session removeOutput:_output];
            [_session stopRunning];
            // close session: (IDO)
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.55 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [self.delegate didFinishScanningBarcodeWithResult:self barcodeValue:detectionString];
            });
            break;
        }
        else
            _label.text = @"Point at SIM ICCID barcode";
    }

    _highlightView.frame = highlightViewRect;
}

@end

File: Photon-Tinker/ScanBarcodeViewController.swift
//
// Created by Raimundas Sakalauskas on 2019-05-02.
// Copyright (c) 2019 Particle. All rights reserved.
//

import Foundation
import AVFoundation

protocol ScanBarcodeViewControllerDelegate: class {
    func didFinishScanningBarcode(withResult scanBarcodeViewController: ScanBarcodeViewController, barcodeValue: String)
    func didCancelScanningBarcode(_ scanBarcodeViewController: ScanBarcodeViewController)
}


class ScanBarcodeViewController: UIViewController, AVCaptureMetadataOutputObjectsDelegate {
    weak var delegate: ScanBarcodeViewControllerDelegate!

    var session: AVCaptureSession!
    var output: AVCaptureMetadataOutput!
    var prevLayer: AVCaptureVideoPreviewLayer!

    var device: AVCaptureDevice?
    var input: AVCaptureDeviceInput?

//    var label: UILabel!
//    var labelBackground: UIView!

    var overlayImageView: UIImageView!

    var flashlightButton: UIButton!
    var flashlightButtonBackground: UIView!

    var cancelButtonBackground: UIView!
    var cancelButton: UIButton!

    var eSeriesSetup: Bool!

    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        overlayImageView = UIImageView(image: UIImage(named: "ImgOverlayGraphic"))
        overlayImageView.translatesAutoresizingMaskIntoConstraints = false
        overlayImageView.image = overlayImageView.image?.withRenderingMode(.alwaysTemplate)
        overlayImageView.tintColor = UIColor.white
        view.addSubview(overlayImageView)

        NSLayoutConstraint.activate([
            self.overlayImageView.centerXAnchor.constraint(equalTo: self.view.centerXAnchor),
            self.overlayImageView.centerYAnchor.constraint(equalTo: self.view.centerYAnchor)
        ])

        if (self.eSeriesSetup) {
            NSLayoutConstraint.activate([
                self.overlayImageView.heightAnchor.constraint(equalTo: self.overlayImageView.widthAnchor)
            ])
        }

        flashlightButtonBackground = UIView(frame: .zero)
        flashlightButtonBackground.translatesAutoresizingMaskIntoConstraints = false
        flashlightButtonBackground.alpha = 0.5
        flashlightButtonBackground.layer.cornerRadius = 20
        flashlightButtonBackground.backgroundColor = UIColor.gray
        view.addSubview(flashlightButtonBackground)

        NSLayoutConstraint.activate([
            self.flashlightButtonBackground.widthAnchor.constraint(equalToConstant: 40),
            self.flashlightButtonBackground.heightAnchor.constraint(equalToConstant: 40)
        ])

        flashlightButton = UIButton(type: .custom)
        flashlightButton.translatesAutoresizingMaskIntoConstraints = false
        flashlightButton.setImage(UIImage(named: "ImgFlashlight"), for: .normal)
        flashlightButton.tintColor = UIColor.white
        flashlightButton.addTarget(self, action: #selector(toggleTorchButton), for: .touchUpInside)
        view.addSubview(flashlightButton)

        NSLayoutConstraint.activate([
            self.flashlightButton.centerXAnchor.constraint(equalTo: self.flashlightButtonBackground.centerXAnchor),
            self.flashlightButton.centerYAnchor.constraint(equalTo: self.flashlightButtonBackground.centerYAnchor),
            self.flashlightButton.widthAnchor.constraint(equalTo: self.flashlightButtonBackground.widthAnchor, constant: -16),
            self.flashlightButton.heightAnchor.constraint(equalTo: self.flashlightButtonBackground.heightAnchor, constant: -16)
        ])

        cancelButtonBackground = UIView(frame: .zero)
        cancelButtonBackground.translatesAutoresizingMaskIntoConstraints = false
        cancelButtonBackground.alpha = 0.5
        cancelButtonBackground.layer.cornerRadius = 20
        cancelButtonBackground.backgroundColor = UIColor.gray
        view.addSubview(cancelButtonBackground)

        NSLayoutConstraint.activate([
            self.cancelButtonBackground.widthAnchor.constraint(equalToConstant: 40),
            self.cancelButtonBackground.heightAnchor.constraint(equalToConstant: 40)
        ])

        cancelButton = UIButton(type: .custom)
        cancelButton.translatesAutoresizingMaskIntoConstraints = false
        cancelButton.setImage(UIImage(named: "IconClose"), for: .normal)
        cancelButton.tintColor = UIColor.white
        cancelButton.addTarget(self, action: #selector(cancelScan), for: .touchUpInside)
        view.addSubview(cancelButton)

        NSLayoutConstraint.activate([
            self.cancelButton.centerXAnchor.constraint(equalTo: self.cancelButtonBackground.centerXAnchor),
            self.cancelButton.centerYAnchor.constraint(equalTo: self.cancelButtonBackground.centerYAnchor),
            self.cancelButton.widthAnchor.constraint(equalTo: self.cancelButtonBackground.widthAnchor, constant: -16),
            self.cancelButton.heightAnchor.constraint(equalTo: self.cancelButtonBackground.heightAnchor, constant: -16)
        ])

        if #available(iOS 11, *) {
            NSLayoutConstraint.activate([
                self.overlayImageView.widthAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.widthAnchor, multiplier: 0.75),

                self.flashlightButtonBackground.rightAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.rightAnchor, constant: -16),
                self.flashlightButtonBackground.topAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.topAnchor, constant: 16),

                self.cancelButtonBackground.leftAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.leftAnchor, constant: 16),
                self.cancelButtonBackground.topAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.topAnchor, constant: 16)
            ])
        } else {
            NSLayoutConstraint.activate([
                self.overlayImageView.widthAnchor.constraint(equalTo: self.view.widthAnchor, multiplier: 0.75),

                self.flashlightButtonBackground.rightAnchor.constraint(equalTo: self.view.rightAnchor, constant: -16),
                self.flashlightButtonBackground.topAnchor.constraint(equalTo: self.view.topAnchor, constant: 36),

                self.cancelButtonBackground.leftAnchor.constraint(equalTo: self.view.leftAnchor, constant: 16),
                self.cancelButtonBackground.topAnchor.constraint(equalTo: self.view.topAnchor, constant: 36)
            ])
        }



        session = AVCaptureSession()
        device = AVCaptureDevice.default(for: .video)

        guard let device = device else {
            //TODO: fix this?...
            fatalError("no AVCaptureDevice")
        }

        input = try? AVCaptureDeviceInput(device: device)
        guard let input = input else {
            //TODO: fix this?...
            fatalError("no AVCaptureDeviceInput")
        }
        session.addInput(input)

        output = AVCaptureMetadataOutput()
        output.setMetadataObjectsDelegate(self, queue: .main)
        session.addOutput(output)

        output.metadataObjectTypes = output.availableMetadataObjectTypes

        prevLayer = AVCaptureVideoPreviewLayer(session: session)
        prevLayer.frame = view.bounds
        prevLayer.videoGravity = .resizeAspectFill
        view.layer.insertSublayer(prevLayer, at: 0)

        session.startRunning()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        if let device = device, device.hasTorch {
            do {
                try device.lockForConfiguration()
                try device.setTorchModeOn(level: 1.0)
                device.unlockForConfiguration()
            } catch {
                print("Torch could not be used")
            }
        }

    }

    @objc func toggleTorchButton(sender: UIButton) {
        if let device = device, device.hasTorch {

            do {
                try device.lockForConfiguration()

                if device.torchMode == .off {
                    try device.setTorchModeOn(level: 1.0)
                } else {
                    device.torchMode = .off
                }

                device.unlockForConfiguration()
            } catch {
                print("Torch could not be used")
            }

        }
    }

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()

        prevLayer.frame = view.bounds
    }


    @objc func cancelScan(_ sender: UIButton) {
        delegate.didCancelScanningBarcode(self)
    }


    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        var detectionString: String? = nil

        for metadata in metadataObjects {
            if (metadata.type == .code128) {
                guard let barCodeObject = prevLayer.transformedMetadataObject(for: metadata) as? AVMetadataMachineReadableCodeObject else { return }
                detectionString = barCodeObject.stringValue

                AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
            } else if (metadata.type == .dataMatrix) {
                AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
                if (self.eSeriesSetup) {
                    guard let readableObject = prevLayer.transformedMetadataObject(for: metadata) as? AVMetadataMachineReadableCodeObject else {
                        return
                    }
                    detectionString = readableObject.stringValue
                } else {
                    session.stopRunning()
                    let ac = UIAlertController(title: TinkerStrings.ScanBarcode.Error.WrongStickerError.Title, message: TinkerStrings.ScanBarcode.Error.WrongStickerError.Message, preferredStyle: .alert)
                    ac.addAction(UIAlertAction(title: TinkerStrings.Action.Ok, style: .default) { [weak self] action in
                        if let self = self {
                            self.session.startRunning()
                        }
                    })
                    present(ac, animated: true)
                }
            }
        }

        if detectionString != nil {
            session.stopRunning()

            DispatchQueue.main.async {
                self.delegate.didFinishScanningBarcode(withResult: self, barcodeValue: detectionString!)
            }
        }
    }

}
